src/runtime/managedtype.cs

```diff
@@ -80,48 +79,41 @@ internal void FreeGCHandle()
         /// </summary>
         internal static ManagedType GetManagedObject(IntPtr ob)
         {
-            if (ob != IntPtr.Zero)
+            if (ob == IntPtr.Zero)
             {
-                IntPtr tp = Runtime.PyObject_TYPE(ob);
-                if (tp == Runtime.PyTypeType || tp == Runtime.PyCLRMetaType)
-                {
-                    tp = ob;
-                }
-
-                var flags = Util.ReadCLong(tp, TypeOffset.tp_flags);
-                if ((flags & TypeFlags.Managed) != 0)
-                {
-                    IntPtr op = tp == ob
-                        ? Marshal.ReadIntPtr(tp, TypeOffset.magic())
-                        : Marshal.ReadIntPtr(ob, ObjectOffset.magic(tp));
-                    if (op == IntPtr.Zero)
-                    {
-                        return null;
-                    }
-                    var gc = (GCHandle)op;
-                    return (ManagedType)gc.Target;
-                }
+                return null;
+            }
+            IntPtr tp = Runtime.PyObject_TYPE(ob);
+            var flags = Util.ReadCLong(tp, TypeOffset.tp_flags);
+            if ((flags & TypeFlags.Managed) == 0)
+            {
+                return null;
+            }
+            int offset = ObjectOffset.magic(tp);
+            if (offset == 0)
+            {
+                return null;
             }
-            return null;
+            IntPtr op = Marshal.ReadIntPtr(ob, offset);
+            if (op == IntPtr.Zero)
+            {
+                return null;
+            }
+            var gc = (GCHandle)op;
+            return (ManagedType)gc.Target;
         }
 
         /// <summary>
         /// Given a Python object, return the associated managed object type or null.
         /// </summary>
         internal static ManagedType GetManagedObjectType(IntPtr ob)
         {
-            if (ob != IntPtr.Zero)
+            if (ob == IntPtr.Zero)
             {
-                IntPtr tp = Runtime.PyObject_TYPE(ob);
-                var flags = Util.ReadCLong(tp, TypeOffset.tp_flags);
-                if ((flags & TypeFlags.Managed) != 0)
-                {
-                    tp = Marshal.ReadIntPtr(tp, TypeOffset.magic());
-                    var gc = (GCHandle)tp;
-                    return (ManagedType)gc.Target;
-                }
+                return null;
             }
-            return null;
+            IntPtr tp = Runtime.PyObject_TYPE(ob);
+            return GetManagedObject(tp);
```

Comment:
I think this code is not equivalent to the original code. The original code did `ob.type.tp_flags.HasFlag(Managed)`, and the new code only does `ob.type.type.tp_flags.HasFlag(Managed)`.
I believe the old check might also need to be necessary.

Reply:
Yes, it's not equivalent. It becomes simple since the memory model been unified.
You can treat the old code was a trick: ob may be a type or an object. You can also see I removed those code `tpHandle == pyHandle  => means it's a ClassBase object`. For now, if the type of `ob` is managed, then the ob must be a managed object, it's obvious on the logic.